Content.Shared/Silicons/StationAi/SharedStationAiSystem.Airlock.cs

```diff
@@ -1,25 +1,117 @@
 
 public abstract partial class SharedStationAiSystem
 {
+    private static readonly SoundPathSpecifier AirlockOverchargeDisabled = new("/Audio/Machines/airlock_overcharge_on.ogg");
+    private static readonly SoundPathSpecifier AirlockOverchargeEnabled = new("/Audio/Machines/airlock_overcharge_off.ogg");
+
     // Handles airlock radial
 
     private void InitializeAirlock()
     {
         SubscribeLocalEvent<DoorBoltComponent, StationAiBoltEvent>(OnAirlockBolt);
+        SubscribeLocalEvent<AirlockComponent, StationAiEmergencyAccessEvent>(OnAirlockEmergencyAccess);
+        SubscribeLocalEvent<ElectrifiedComponent, StationAiElectrifiedEvent>(OnElectrified);
     }
 
+    /// <summary>
+    /// Attempts to bolt door. If wire was cut (AI or for bolts) or its not powered - notifies AI and does nothing.
+    /// </summary>
     private void OnAirlockBolt(EntityUid ent, DoorBoltComponent component, StationAiBoltEvent args)
     {
+        if (!TryComp<StationAiWhitelistComponent>(ent, out var whiteList))
+        {
+            return;
+        }
+
+        if (!whiteList.Enabled || component.BoltWireCut || !component.Powered)
+        {
+            _popup.PopupClient(Loc.GetString("ai-device-not-responding"), args.User, PopupType.MediumCaution);
+            return;
+        }
+
         _doors.SetBoltsDown((ent, component), args.Bolted, args.User, predicted: true);
     }
+
+    /// <summary>
+    /// Attempts to bolt door. If wire was cut (AI) or its not powered - notifies AI and does nothing.
+    /// </summary>
+    private void OnAirlockEmergencyAccess(EntityUid ent, AirlockComponent component, StationAiEmergencyAccessEvent args)
+    {
+        if (!TryComp<StationAiWhitelistComponent>(ent, out var whiteList))
+        {
+            return;
+        }
+
+        if (!whiteList.Enabled || !component.Powered)
+        {
+            _popup.PopupClient(Loc.GetString("ai-device-not-responding"), args.User, PopupType.MediumCaution);
+            return;
+        }
+
+        _airlocks.SetEmergencyAccess((ent, component), args.EmergencyAccess, args.User, predicted: true);
+    }
+
+    /// <summary>
+    /// Attempts to bolt door. If wire was cut (AI or for one of power-wires) or its not powered - notifies AI and does nothing.
+    /// </summary>
+    private void OnElectrified(EntityUid ent, ElectrifiedComponent component, StationAiElectrifiedEvent args)
+    {
+        if (!TryComp<ElectrifiedComponent>(ent, out var electrified)
+            || !TryComp<StationAiWhitelistComponent>(ent, out var whiteList))
+        {
+            return;
+        }
+
+        SharedApcPowerReceiverComponent? apcPowerReceiverComponent = null;
+        if (
+            !whiteList.Enabled
+            || electrified.IsWireCut
+            || (
+                PowerReceiver.ResolveApc(ent, ref apcPowerReceiverComponent)
+                && !apcPowerReceiverComponent.Powered
+                )
+        )
+        {
```

Comment:
Whitelist comp gets handled in the BUI attempt event, having to check this manually every single time is bug prone and shouldn't be done.

Just add the whitelist enabled check to it.

Reply:
i'll send popup from there, ye. It checks out pretty well.

Reply:
fixed